src/data_api/events.se

```diff
@@ -1,346 +1,420 @@
+    self.events[eventId].tokens[branch][outcome][sourceAddress] = safeSub(startingSourceAttotoken, attotoken)
+    self.events[eventId].tokens[branch][outcome][destinationAddress] = safeAdd(startingDestinationAttotoken, attotoken)
+    log(type = Transfer, eventId, branch, outcome, sourceAddress, destinationAddress, attotoken)
     return(1)
 
-def setRejected(event, period):
-    self.controller.assertIsWhitelisted(msg.sender)
-    self.Events[event].rejected = 1
-    self.Events[event].rejectedPeriod = period
+def privateBuyToken(eventId, outcome, reporterAddress, attotoken):
+    privateOnly()
+    # transfer REP from user to events contract
+    reputationContract = self.controller.lookup('reputation')
+    branch = self.events[eventId].branch
+    # reputationContract.transferFrom(branch, reporterAddress, self, attorepStake)
+    # create/assign outcome token
+    startingAttotoken = self.events[eventId].tokens[branch][outcome][reporterAddress]
+    startingTokenSupply = self.events[eventId].tokenSupply[branch][outcome]
+    startingAllOutcomesTokenSupply = self.events[eventId].allOutcomesTokenSupply[branch]
+    self.events[eventId].tokens[branch][outcome][reporterAddress] = safeAdd(startingAttotoken, attotoken)
+    self.events[eventId].tokenSupply[branch][outcome] = safeAdd(startingTokenSupply, attotoken)
+    self.events[eventId].allOutcomesTokenSupply[branch] = safeAdd(startingAllOutcomesTokenSupply, attotoken)
+
+# if another event caused a fork after users had bought outcome tokens on this branch, allow them to be traded back in for REP at face value on this branch
+def privateTryRedeemTokensForNonForkingEventOnForkedBranch(eventId, branch, outcome, tokenHolderAddress):
+    privateOnly()
+    forkingEventId = self.controller.lookup('branches').getForkingEvent(branch)
+    if (forkingEventId == 0):
+        return(0)
+    if (forkingEventId == eventId):
+        return(0)
+    tokensToRedeem = self.events[eventId].tokens[branch][outcome][tokenHolderAddress]
+    self.events[eventId].tokens[branch][outcome][tokenHolderAddress] = 0
+    # self.controller.lookup('reputationToken').transfer(branch, tokenHolderAddress, tokensToRedeem)
     return(1)
 
-def getRejected(event):
-    return(self.Events[event].rejected)
-
-def getRejectedPeriod(event):
-    return(self.Events[event].rejectedPeriod)
-
-def setExtraBondPoster(event, poster):
-    self.controller.assertIsWhitelisted(msg.sender)
-    self.Events[event].extraBondPoster = poster
+# if this event is the cause of a fork, force tokens to be redeemed on the branch associated with the token's outcome
+def privateTryRedeemTokensOnForkingEvent(eventId, branch, outcome, tokenHolderAddress):
+    privateOnly()
+    # FIXME: need to redeem fees, but the current implementation just redeems tokens on outcome branch. CONSIDER: change this to only allow token redemption after event is finalized, then we can also distribute fees on correct outcome branch
```

Comment:
Reminder to consider failed bonds [ones where the final outcome was no different than the outcome when posted] in the rep redistribution here as well --- this is in the bottom section of the diagram